// @coxautokc/fusion-ui-components/lib/ActionItem/ActionItem.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * ActionItem is used as a wrapper for each ActionBar child to validate props.<br/>
 * It should be accessed as ActionBar.Item as it should only be used in conjunction with ActionBar.
 */
var ActionItem = function ActionItem() {
  return _react2.default.createElement(
    'div',
    null,
    'This is an empty component, only used to verify props for ActionBar children'
  );
};

ActionItem.propTypes = {
  /** Type of action taken when ActionItem is clicked.
   * <ul>
   *  <li><strong>collapsiblePanel</strong>: Drawer behavior only available when actionTab is set to staticTab.</li>
   *  <li><strong>modal</strong>: Opens a modal.</li>
   *  <li><strong>link</strong>: Links to another page using link prop.</li>
   *  <li><strong>none</strong>: Performs no action. Useful when tab's behavior is configured elsewhere.</li>
   * </ul>
  */
  actionDisplay: _propTypes2.default.oneOf(['collapsiblePanel', 'modal', 'link', 'none']).isRequired,

  /** Specifies whether action item is a tab, dropdown, or fluid.<br />
   * <ul>
   *  <li><strong>staticTab</strong>: Always render tab.</li>
   *  <li><strong>staticDropdown</strong>: Always render dropdown.</li>
   *  <li><strong>fluid</strong>: Render as tab initially. Overflow to dropdown item if needed.</li>
   * </ul>
  */
  actionPosition: _propTypes2.default.oneOf(['staticTab', 'staticDropdown', 'fluid']).isRequired,

  /** Set to false to disable the ActionItem. */
  enable: _propTypes2.default.bool.isRequired,

  /** Icon for ActionItem. */
  icon: _propTypes2.default.node,

  /** Label for ActionItem. */
  label: _propTypes2.default.string.isRequired,

  /** Link for ActionItem. Accepts all parameters for <a href="https://developer.mozilla.org/en-US/docs/Web/API/Window/open">window.open()</a>. */
  link: _propTypes2.default.shape({
    url: _propTypes2.default.string,
    target: _propTypes2.default.string,
    specs: _propTypes2.default.string,
    replace: _propTypes2.default.bool
  }),

  /** Static identifier to use as reference for each ActionItem. */
  name: _propTypes2.default.string,

  /**  External function to fire onClick. */
  onClick: _propTypes2.default.func,

  /** Component to open onClick.
   * If supplied a changeHandler, onClick will fire it instead.
  */
  openToComponent: _propTypes2.default.oneOfType([_propTypes2.default.node, _propTypes2.default.func]),

  /** Classes given to tabbed ActionItems to control icon and link visibility. */
  responsiveClasses: _propTypes2.default.object,

  /** Set to false to hide ActionItem. */
  show: _propTypes2.default.bool,

  /** If set to true, show divider after tab or dropdown item. */
  showDivider: _propTypes2.default.bool.isRequired
};

ActionItem.defaultProps = {
  actionPosition: 'fluid',
  actionDisplay: 'none',
  enable: true,
  link: {
    target: '_blank',
    replace: false
  },
  show: true,
  showDivider: false
};

exports.default = ActionItem;
