// @coxautokc/fusion-ui-components/lib/internal/AccordionItem.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Panel = require('react-bootstrap/lib/Panel');

var _Panel2 = _interopRequireDefault(_Panel);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

/** AccordionItem is merely a wrapper for <Panel> when used in conjuction with React-Bootstrap's <Accordion> component.
    AccordionItem accepts all Panel properties, except for "bsClass".
    See https://react-bootstrap.github.io/components.html#panels-accordion for implementation of Accordion.
*/
var AccordionItem = function AccordionItem(props) {
  var htmlId = props.htmlId,
      header = props.header,
      bodyRender = props.bodyRender,
      children = props.children,
      className = props.className,
      other = _objectWithoutProperties(props, ['htmlId', 'header', 'bodyRender', 'children', 'className']);

  var accordionHeader = props.header ? header : ' ';
  var mergedClasses = (className ? className : '') + (!props.header ? 'accordion--no-heading' : '') + (props.bodyRender ? 'accordion--body-hide' : '');

  return _react2.default.createElement(
    _Panel2.default,
    Object.assign({
      bsClass: 'accordion',
      className: mergedClasses,
      header: accordionHeader,
      id: htmlId
    }, other),
    children
  );
};

AccordionItem.propTypes = {
  bodyRender: _propTypes2.default.bool, // Accordion body content has display:none by default, causing dynamic contnet (like, FusionCharts) to no render since it's container is hidden.  This allows content to render, but adds overflow:hidden;
  children: _propTypes2.default.node,
  className: _propTypes2.default.string,
  header: _propTypes2.default.node,
  htmlId: _propTypes2.default.string.isRequired
};

exports.default = AccordionItem;
