// @coxautokc/fusion-ui-components/lib/internal/AxisLabel.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var AxisLabel = function AxisLabel(_ref) {
  var w = _ref.w,
      h = _ref.h,
      axisLabel = _ref.axisLabel,
      axisType = _ref.axisType,
      padding = _ref.padding;

  var translateLabelX = "translate(" + w / 2 + "," + (h + 40) + ")";
  var translateLabelY = "translate(" + (-40 - padding) + "," + h / 2 + ") rotate(270)";

  return _react2.default.createElement(
    'text',
    { className: axisType + "-axis-label", textAnchor: 'middle', transform: axisType === 'y' ? translateLabelY : translateLabelX },
    axisLabel
  );
};

AxisLabel.propTypes = {
  axisLabel: _propTypes2.default.string,
  axisType: _propTypes2.default.oneOf(['x', 'y']),
  h: _propTypes2.default.number,
  padding: _propTypes2.default.number,
  w: _propTypes2.default.number
};

AxisLabel.defaultProps = {
  padding: 0
};

exports.default = AxisLabel;
