// @coxautokc/fusion-ui-components/lib/internal/Grid.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _d = require('d3');

var _d2 = _interopRequireDefault(_d);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Grid = function (_React$Component) {
  _inherits(Grid, _React$Component);

  function Grid() {
    _classCallCheck(this, Grid);

    return _possibleConstructorReturn(this, (Grid.__proto__ || Object.getPrototypeOf(Grid)).apply(this, arguments));
  }

  _createClass(Grid, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.renderGrid();
    }
  }, {
    key: 'componentDidUpdate',
    value: function componentDidUpdate() {
      this.renderGrid();
    }
  }, {
    key: 'renderGrid',
    value: function renderGrid() {
      var node = _reactDom2.default.findDOMNode(this);
      _d2.default.select(node).call(this.props.grid);
    }
  }, {
    key: 'render',
    value: function render() {
      var translate = "translate(0," + this.props.h + ")";
      return _react2.default.createElement('g', { className: 'y-grid', stroke: this.props.lineColor, strokeWidth: this.props.strokeWidth, transform: this.props.gridType === 'x' ? translate : "" });
    }
  }]);

  return Grid;
}(_react2.default.Component);

Grid.propTypes = {
  grid: _propTypes2.default.func,
  gridType: _propTypes2.default.oneOf(['x', 'y']),
  h: _propTypes2.default.number,
  lineColor: _propTypes2.default.string,
  strokeWidth: _propTypes2.default.number
};

Grid.defaultProps = {
  lineColor: '#ccc',
  strokeWidth: 1
};

exports.default = Grid;
