// @coxautokc/fusion-ui-components/lib/PhoneInput/PhoneInput.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _MaskedInput = require('../internal/MaskedInput');

var _MaskedInput2 = _interopRequireDefault(_MaskedInput);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * PhoneInput provides a phone masked and validated wrapper to TextInput.
 * It includes the full Form group including
 * the Label, Control, and helpText. PhoneInput handles its own
 * required validation but can be overridden by passing in error on
 * props. PhoneInput will pass back both value (unmasked) and maskedValue on event.target.
 */

var PhoneInput = function (_React$Component) {
  _inherits(PhoneInput, _React$Component);

  function PhoneInput(props) {
    _classCallCheck(this, PhoneInput);

    var _this = _possibleConstructorReturn(this, (PhoneInput.__proto__ || Object.getPrototypeOf(PhoneInput)).call(this, props));

    _this.getWarningMessage = function () {
      var value = _this.unmask(_this.props.value);
      return _this.isPhoneNumberValid(value) ? '' : 'Received invalid phone number';
    };

    _this.unmask = function (value) {
      return value ? value.replace(/[ ()_-]/g, '') : value;
    };

    _this.onChange = function (maskedValue, isValid) {
      var value = _this.unmask(maskedValue);

      var error = !_this.isPhoneNumberValid(value) ? _this.props.label + " is Invalid" : "";
      var isFieldValid = isValid && error.length <= 0;

      _this.props.onChange({ target: { name: _this.props.name, value: value, maskedValue: maskedValue } }, isFieldValid);
      _this.setState({ error: error });
    };

    _this.isPhoneNumberValid = function (value) {
      return !(value.length < 10 && value.length > 0);
    };

    _this.mask = ['(', /[1-9]/, /\d/, /\d/, ')', ' ', /\d/, /\d/, /\d/, '-', /\d/, /\d/, /\d/, /\d/];
    _this.state = {
      error: _this.props.error || '',
      warning: _this.getWarningMessage()
    };
    return _this;
  }

  _createClass(PhoneInput, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          className = _props.className,
          displayLabel = _props.displayLabel,
          htmlId = _props.htmlId,
          label = _props.label,
          name = _props.name,
          placeholder = _props.placeholder,
          required = _props.required,
          value = _props.value,
          props = _objectWithoutProperties(_props, ['children', 'className', 'displayLabel', 'htmlId', 'label', 'name', 'placeholder', 'required', 'value']);

      var _state = this.state,
          error = _state.error,
          warning = _state.warning;

      return _react2.default.createElement(
        _MaskedInput2.default,
        Object.assign({
          className: className,
          displayLabel: displayLabel,
          error: error,
          warning: warning,
          htmlId: htmlId,
          label: label,
          mask: this.mask,
          maxLength: 15,
          name: name,
          onMaskedInputChange: this.onChange,
          placeholder: placeholder,
          required: required,
          value: value
        }, props),
        children
      );
    }
  }]);

  return PhoneInput;
}(_react2.default.Component);

PhoneInput.propTypes = {
  /** Any children to be passed through. */
  children: _propTypes2.default.node,

  /** Additional class name. */
  className: _propTypes2.default.string,

  /** Boolean displays label if set to true. */
  displayLabel: _propTypes2.default.bool,

  /** If error is coming from parent component,
   * use this prop to pass the error message.
   * This prop will overwrite the 'PhoneNumber is required' and
   * 'PhoneNumber is Invalid' errors untill it is cleared.   */
  error: _propTypes2.default.string,

  /** Globally unique and descriptive HTML ID. Used by QA for automated testing. */
  htmlId: _propTypes2.default.string.isRequired,

  /** Input label. */
  label: _propTypes2.default.string.isRequired,

  /** Input name. */
  name: _propTypes2.default.string.isRequired,

  /** onChange callback. Called with <code>cb(event, isValid)</code>. Check isValid to determine if the control is in error. */
  onChange: _propTypes2.default.func.isRequired,

  /** Function to call onFocus. */
  onFocus: _propTypes2.default.func,

  /** Placeholder to display when empty. */
  placeholder: _propTypes2.default.string,

  /** Mark label with asterisk and perform
   * required field validation on blur if set to true. */
  required: _propTypes2.default.bool,

  /** Text input value. */
  value: _propTypes2.default.string.isRequired
};

exports.default = PhoneInput;
