// @coxautokc/fusion-ui-components/lib/Col/Col.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Col = require('react-bootstrap/lib/Col');

var _Col2 = _interopRequireDefault(_Col);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

/** Grid framework for responsive content */
var Col = function Col(_ref) {
  var children = _ref.children,
      htmlId = _ref.htmlId,
      props = _objectWithoutProperties(_ref, ['children', 'htmlId']);

  delete props.bsClass;

  return _react2.default.createElement(
    _Col2.default,
    Object.assign({ id: htmlId }, props),
    children
  );
};

Col.propTypes = {
  /** Base CSS class and prefix for the component. */
  bsClass: _propTypes2.default.string,

  /** Any child components. */
  children: _propTypes2.default.node,

  /** Component className. */
  className: _propTypes2.default.string,

  /** You can use a custom element for this component. */
  componentClass: _propTypes2.default.string,

  /** Globally unique and descriptive HTML ID. Used by QA for automated testing. */
  htmlId: _propTypes2.default.string,

  /** The number of columns you wish to span for Large devices Desktops (≥1200px) </br> class-prefix <code>col-lg-</code> */
  lg: _propTypes2.default.number,

  /** Hide column on Large devices Desktops </br> adds class <code>hidden-lg</code> */
  lgHidden: _propTypes2.default.bool,

  /** Move columns to the right for Large devices Desktops </br> class-prefix <code>col-lg-offset-</code> */
  lgOffset: _propTypes2.default.number,

  /** Change the order of grid columns to the left for Large devices Desktops </br> class-prefix <code>col-lg-pull-</code> */
  lgPull: _propTypes2.default.number,

  /** Change the order of grid columns to the right for Large devices Desktops </br> class-prefix <code>col-lg-push-</code> */
  lgPush: _propTypes2.default.number,

  /** The number of columns you wish to span for Medium devices Desktops (≥992px) </br> class-prefix <code>col-md-</code> */
  md: _propTypes2.default.number,

  /** Hide column on Medium devices Desktops </br> adds class <code>hidden-md</code> */
  mdHidden: _propTypes2.default.bool,

  /** Move columns to the right for Medium devices Desktops </br> class-prefix <code>col-md-offset-</code> */
  mdOffset: _propTypes2.default.number,

  /** Change the order of grid columns to the left for Medium devices Desktops </br> class-prefix <code>col-md-pull-</code> */
  mdPull: _propTypes2.default.number,

  /** Change the order of grid columns to the right for Medium devices Desktops </br> class-prefix <code>col-md-push-</code> */
  mdPush: _propTypes2.default.number,

  /** The number of columns you wish to span for Small devices Tablets (≥768px) </br> class-prefix <code>col-sm-</code> */
  sm: _propTypes2.default.number,

  /** Hide column on Small devices Tablets </br> adds class <code>hidden-sm</code> */
  smHidden: _propTypes2.default.bool,

  /** Move columns to the right for Small devices Tablets </br> class-prefix <code>col-sm-offset-</code> */
  smOffset: _propTypes2.default.number,

  /** Change the order of grid columns to the left for Small devices Tablets </br> class-prefix <code>col-sm-pull-</code> */
  smPull: _propTypes2.default.number,

  /** Change the order of grid columns to the right for Small devices Tablets </br> class-prefix <code>col-sm-push-</code> */
  smPush: _propTypes2.default.number,

  /** The number of columns you wish to span for Extra small devices Phones (<768px) </br> class-prefix <code>col-xs-</code> */
  xs: _propTypes2.default.number,

  /** Hide column on Extra small devices Phones </br> adds class <code>hidden-xs</code> */
  xsHidden: _propTypes2.default.bool,

  /** Move columns to the right for Extra small devices Phones </br> class-prefix <code>col-xs-offset-</code> */
  xsOffset: _propTypes2.default.number,

  /** Change the order of grid columns to the left for Extra small devices Phones </br> class-prefix <code>col-xs-pull-</code> */
  xsPull: _propTypes2.default.number,

  /** Change the order of grid columns to the right for Extra small devices Phones </br> class-prefix <code>col-xs-push-</code> */
  xsPush: _propTypes2.default.number
};

Col.defaultProps = {
  componentClass: 'div'
};

exports.default = Col;
