// @coxautokc/taxes/build/components/customizedUiComponents/dropdown.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactSelect = require("react-select");

var _reactSelect2 = _interopRequireDefault(_reactSelect);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Dropdown = function Dropdown(props) {
  var name = props.name,
      label = props.label,
      required = props.required;

  if (!required) {
    return _react2.default.createElement(
      "div",
      { className: "col-xs-12 form-group-taxes" },
      label && _react2.default.createElement(
        "label",
        { htmlFor: name },
        label
      ),
      _react2.default.createElement(_reactSelect2.default, props)
    );
  }

  return _react2.default.createElement(
    "div",
    null,
    _react2.default.createElement(
      "div",
      { className: "col-xs-10 form-group-taxes" },
      label && _react2.default.createElement(
        "label",
        { htmlFor: name },
        label
      ),
      _react2.default.createElement(_reactSelect2.default, props)
    )
  );
};

Dropdown.propTypes = {
  disabled: _propTypes2.default.bool,
  label: _propTypes2.default.string,
  name: _propTypes2.default.string.isRequired,
  onChange: _propTypes2.default.func.isRequired,
  options: _propTypes2.default.array.isRequired,
  required: _propTypes2.default.bool
};

exports.default = Dropdown;
