// @coxautokc/fusion-ui-components/lib/NumericInput/NumericInput.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _MaskedInput = require('../internal/MaskedInput');

var _MaskedInput2 = _interopRequireDefault(_MaskedInput);

var _createNumberMask = require('text-mask-addons/dist/createNumberMask');

var _createNumberMask2 = _interopRequireDefault(_createNumberMask);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/** NumericInput uses a masked input and only allows numerical values.
 * It wraps the full Form group including the Label, Control, and helpText.
 * NumericInput will not allow decimals or negative numbers by default,
 * but can if allowDecimal and allowNegative props are set to true.
*/

var NumericInput = function (_React$Component) {
  _inherits(NumericInput, _React$Component);

  function NumericInput(props) {
    _classCallCheck(this, NumericInput);

    var _this = _possibleConstructorReturn(this, (NumericInput.__proto__ || Object.getPrototypeOf(NumericInput)).call(this, props));

    _this.getMask = function () {
      return (0, _createNumberMask2.default)({
        prefix: '',
        suffix: '',
        includeThousandsSeparator: _this.props.autoInsertCommas,
        thousandsSeparatorSymbol: ',',
        allowDecimal: _this.props.allowDecimal,
        decimalSymbol: '.',
        decimalLimit: null,
        integerLimit: null,
        requireDecimal: false,
        allowNegative: _this.props.allowNegative,
        allowLeadingZeroes: true
      });
    };

    _this.unmask = function (value) {
      return value ? value.replace(/,/g, '') : value;
    };

    _this.onChange = function (maskedValue, isValid) {
      var value = _this.unmask(maskedValue);

      _this.props.onChange({ target: { name: _this.props.name, value: value, maskedValue: maskedValue } }, isValid);
      _this.setState({ value: value });
    };

    _this.mask = _this.getMask();
    _this.state = {
      error: _this.props.error || '',
      value: _this.props.value || ''
    };
    return _this;
  }

  _createClass(NumericInput, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          className = _props.className,
          displayLabel = _props.displayLabel,
          htmlId = _props.htmlId,
          label = _props.label,
          name = _props.name,
          placeholder = _props.placeholder,
          required = _props.required,
          minLength = _props.minLength,
          maxLength = _props.maxLength,
          disabled = _props.disabled;

      return _react2.default.createElement(_MaskedInput2.default, {
        className: className,
        displayLabel: displayLabel,
        error: this.state.error,
        htmlId: htmlId,
        label: label,
        mask: this.mask,
        maxLength: maxLength,
        minLength: minLength,
        name: name,
        required: required,
        onMaskedInputChange: this.onChange,
        placeholder: placeholder,
        value: this.state.value,
        disabled: disabled
      });
    }
  }]);

  return NumericInput;
}(_react2.default.Component);

NumericInput.propTypes = {
  /** Allows decimals if set to true. */
  allowDecimal: _propTypes2.default.bool,

  /** Allows negative value if set to true. */
  allowNegative: _propTypes2.default.bool,

  /** Automatically inserts commas as thousand separators if set to true */
  autoInsertCommas: _propTypes2.default.bool,

  /** Additional class name. */
  className: _propTypes2.default.string,

  /** Set to true to disable the NumericInput. */
  disabled: _propTypes2.default.bool,

  /** Display label if set to true. */
  displayLabel: _propTypes2.default.bool,

  /** String to display when error occurs. */
  error: _propTypes2.default.string,

  /** Globally unique and descriptive HTML ID. Used by QA for automated testing. */
  htmlId: _propTypes2.default.string.isRequired,

  /** Input label. */
  label: _propTypes2.default.string.isRequired,

  /** Maximum length of characters allowed. */
  maxLength: _propTypes2.default.number,

  /** Minimum number of characters required. */
  minLength: _propTypes2.default.number,

  /** Input name. */
  name: _propTypes2.default.string.isRequired,

  /** onChange callback. Called with <code>cb(event, isValid)</code>. Check isValid to determine if the control is in error. */
  onChange: _propTypes2.default.func.isRequired,

  /** Function to call onFocus. */
  onFocus: _propTypes2.default.func,

  /** Placeholder to display when empty. */
  placeholder: _propTypes2.default.string,

  /** Mark label with asterisk and perform required field validation on blur if set to true. */
  required: _propTypes2.default.bool,

  /** Text input value. */
  value: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]).isRequired

};

NumericInput.defaultProps = {
  autoInsertCommas: false,
  allowDecimal: false,
  allowNegative: false,
  displayLabel: true
};

exports.default = NumericInput;
