// @coxautokc/fusion-ui-components/lib/ActionBar/ActionDrawer.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ActionDrawer = function ActionDrawer(_ref) {
  var htmlId = _ref.htmlId,
      actionComponent = _ref.actionComponent,
      closePanel = _ref.closePanel,
      showActionPanel = _ref.showActionPanel;

  // cloneElement is used to add ActionBar's props to the Component
  // This specifically passes down the toggle function from this file
  return _react2.default.cloneElement(actionComponent, {
    htmlId: htmlId,
    closePanel: closePanel,
    showActionPanel: showActionPanel
  });
};

ActionDrawer.propTypes = {
  actionComponent: _propTypes2.default.object.isRequired,
  closePanel: _propTypes2.default.func.isRequired,
  htmlId: _propTypes2.default.string.isRequired,
  showActionPanel: _propTypes2.default.bool.isRequired
};

exports.default = ActionDrawer;
