// @coxautokc/fusion-ui-components/lib/ActionBar/ActionDialog.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ActionDialog = function ActionDialog(_ref) {
  var htmlId = _ref.htmlId,
      actionComponent = _ref.actionComponent,
      closeDialog = _ref.closeDialog,
      showActionDialog = _ref.showActionDialog;

  // cloneElement is used to add ActionBar's props to the Component
  // This specifically passes down the toggle function from this file
  return _react2.default.cloneElement(actionComponent, {
    htmlId: htmlId,
    closeDialog: closeDialog,
    showActionDialog: showActionDialog
  });
};

ActionDialog.propTypes = {
  actionComponent: _propTypes2.default.object.isRequired,
  closeDialog: _propTypes2.default.func.isRequired,
  htmlId: _propTypes2.default.string.isRequired,
  showActionDialog: _propTypes2.default.bool.isRequired
};

exports.default = ActionDialog;
