// @coxautokc/rebates/build/components/RebatesFormElements/RebatesFormElements.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Input = require("../customizedUIComponents/Input/Input");

var _Input2 = _interopRequireDefault(_Input);

var _Row = require("@coxautokc/fusion-ui-components/lib/Row");

var _Row2 = _interopRequireDefault(_Row);

var _Col = require("@coxautokc/fusion-ui-components/lib/Col");

var _Col2 = _interopRequireDefault(_Col);

require("./RebatesFormElements.scss");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var RebatesFormElements = function (_React$Component) {
  _inherits(RebatesFormElements, _React$Component);

  function RebatesFormElements() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, RebatesFormElements);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = RebatesFormElements.__proto__ || Object.getPrototypeOf(RebatesFormElements)).call.apply(_ref, [this].concat(args))), _this), _this.generateOnChangeHandler = function (r, type, position) {
      if (type !== "DealerRebate") {
        return function (e) {
          _this.props.onChange(type, e.target.value);
        };
      }

      return function (e) {
        _this.props.onChange(type, e.target.value, position);
      };
    }, _this.generateOnBlurHandler = function (r, type, position) {
      if (type !== "DealerRebate") {
        return function (e) {
          _this.props.onBlur(type, e.target.value);
        };
      }

      return function (e) {
        _this.props.onBlur(type, e.target.value, position);
      };
    }, _this.generateOnDealerRebateRemoveHandler = function (position) {
      return function () {
        _this.props.onDealerRebateRemove(position);
      };
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(RebatesFormElements, [{
    key: "renderRebateField",
    value: function renderRebateField(r, type, position) {
      return _react2.default.createElement(
        _Row2.default,
        { key: "rebates-row-" + type + (position ? "-" + position : "") },
        type !== "DealerRebate" ? _react2.default.createElement(
          _Col2.default,
          { md: 4, xs: 4 },
          _react2.default.createElement(
            "label",
            { htmlFor: r.type },
            r.name
          )
        ) : "",
        _react2.default.createElement(
          _Col2.default,
          {
            xs: type !== "DealerRebate" ? 5 : 12,
            md: type !== "DealerRebate" ? 5 : 12
          },
          _react2.default.createElement(_Input2.default, {
            htmlId: type + (position ? '-' + position + '-' : '-') + 'Input',
            name: type + (position ? '-' + position + '-' : '-') + 'Input',
            disabled: type !== "DealerRebate",
            addOnIndentation: "left",
            addOnClassName: "fa fa-usd",
            value: r.amount,
            placeholder: "0.00",
            onChange: this.generateOnChangeHandler(r, type, position),
            onBlur: this.generateOnBlurHandler(r, type, position),
            error: this.props.showErrors && r.error
          })
        )
      );
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          rebates = _props.rebates,
          onDealerRebateAdd = _props.onDealerRebateAdd,
          showErrors = _props.showErrors;

      var nonDealerRebateFields = rebates.filter(function (r) {
        return r.type !== "DealerRebate";
      });
      var dealerRebateFields = rebates.find(function (r) {
        return r.type === "DealerRebate";
      }).subElements;
      var maxDealerRebateCountReached = dealerRebateFields.length >= 4;
      var minDealerRebateCountReached = dealerRebateFields.length <= 1;

      return _react2.default.createElement(
        "div",
        { id: "RebatesFormElements" },
        nonDealerRebateFields.map(function (r) {
          return _this2.renderRebateField(r, r.type);
        }),
        _react2.default.createElement(
          _Row2.default,
          null,
          _react2.default.createElement(
            _Col2.default,
            { md: 4, xs: 4 },
            _react2.default.createElement(
              "label",
              { htmlFor: "dealerRebate" },
              "Dealer Rebate"
            )
          ),
          _react2.default.createElement(
            _Col2.default,
            { md: 5, xs: 5 },
            dealerRebateFields.map(function (r, index) {
              return _this2.renderRebateField(r, "DealerRebate", index + 1);
            })
          ),
          _react2.default.createElement(
            _Col2.default,
            { md: 3, xs: 3 },
            dealerRebateFields.map(function (r, index) {
              return _react2.default.createElement(
                _Row2.default,
                { key: index, className: "extra-padded-row" },
                _react2.default.createElement(
                  _Col2.default,
                  { xs: 6, md: 6, className: "dealer-rebate-add" },
                  _react2.default.createElement(
                    "a",
                    {
                      id: "dealer-rebate-add" + "-" + (index + 1),
                      className: maxDealerRebateCountReached || index + 1 !== dealerRebateFields.length ? 'disabled' : '',
                      disabled: maxDealerRebateCountReached || index + 1 !== dealerRebateFields.length,
                      onClick: !(maxDealerRebateCountReached || index + 1 !== dealerRebateFields.length) && onDealerRebateAdd },
                    _react2.default.createElement("i", { className: "fa fa-plus" })
                  )
                ),
                _react2.default.createElement(
                  _Col2.default,
                  { xs: 6, md: 6, className: "dealer-rebate-remove" },
                  _react2.default.createElement(
                    "a",
                    {
                      id: "dealer-rebate-remove" + "-" + (index + 1),
                      className: minDealerRebateCountReached ? 'disabled' : '',
                      disabled: minDealerRebateCountReached,
                      onClick: !minDealerRebateCountReached && _this2.generateOnDealerRebateRemoveHandler(index + 1) },
                    _react2.default.createElement("i", { className: "fa fa-minus" })
                  )
                )
              );
            })
          )
        ),
        showErrors && _react2.default.createElement(
          _Row2.default,
          null,
          _react2.default.createElement(
            _Col2.default,
            { xs: 12, md: 12, className: "error" },
            "* The value must be from $0 to $99,999.99"
          )
        )
      );
    }
  }]);

  return RebatesFormElements;
}(_react2.default.Component);

exports.default = RebatesFormElements;


RebatesFormElements.propTypes = {
  showErrors: _propTypes2.default.bool,
  rebates: _propTypes2.default.array.isRequired,
  onChange: _propTypes2.default.func.isRequired,
  onBlur: _propTypes2.default.func.isRequired,
  onDealerRebateAdd: _propTypes2.default.func.isRequired,
  onDealerRebateRemove: _propTypes2.default.func.isRequired
};
