// @coxautokc/fusion-ui-components/lib/Grid/Grid.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Grid = require('react-bootstrap/lib/Grid');

var _Grid2 = _interopRequireDefault(_Grid);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

/** Grid framework for responsive content */
var Grid = function Grid(_ref) {
  var htmlId = _ref.htmlId,
      children = _ref.children,
      props = _objectWithoutProperties(_ref, ['htmlId', 'children']);

  delete props.bsClass;

  return _react2.default.createElement(
    _Grid2.default,
    Object.assign({ id: htmlId }, props),
    children
  );
};

Grid.propTypes = {
  /** Base CSS class and prefix for the component. */
  bsClass: _propTypes2.default.string,

  /** Any child components. */
  children: _propTypes2.default.node,

  /** Component className. */
  className: _propTypes2.default.string,

  /** You can use a custom element for this component. */
  componentClass: _propTypes2.default.string,

  /** Turn any fixed-width grid layout into a full-width layout by this property. Adds <code>container-fluid</code> class.*/
  fluid: _propTypes2.default.bool,

  /** Globally unique and descriptive HTML ID. Used by QA for automated testing. */
  htmlId: _propTypes2.default.string
};

Grid.defaultProps = {
  componentClass: 'div',
  fluid: true
};

exports.default = Grid;
