// @coxautokc/fusion-ui-components/lib/DataTable/DataTablePagination.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Pagination = require('react-bootstrap/lib/Pagination');

var _Pagination2 = _interopRequireDefault(_Pagination);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var DataTablePagination = function DataTablePagination(_ref) {
  var props = _objectWithoutProperties(_ref, []);

  return _react2.default.createElement(_Pagination2.default, {
    bsSize: 'medium',
    items: props.maxPages,
    activePage: props.currentPage,
    onSelect: props.setPage,
    maxButtons: 5,
    prev: true,
    next: true,
    first: true,
    last: true,
    ellipsis: true,
    boundaryLinks: true
  });
};

DataTablePagination.propTypes = {
  currentPage: _propTypes2.default.number,
  maxPages: _propTypes2.default.number,
  setPage: _propTypes2.default.func
};

exports.default = DataTablePagination;
