// @coxautokc/fees/build/actions/actionTypes.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
// Loadings
var LOADING = exports.LOADING = 'LOADING';
var LOADING_COMPLETE = exports.LOADING_COMPLETE = 'LOADING_COMPLETE';

// Loading Indicator
var LOADING_INDICATOR_START = exports.LOADING_INDICATOR_START = 'LOADING_INDICATOR_START';
var LOADING_INDICATOR_SET_PERCENT = exports.LOADING_INDICATOR_SET_PERCENT = 'LOADING_INDICATOR_SET_PERCENT';

var FETCH_EXISTING_FEE_QUOTE_FEE_SUCCESS = exports.FETCH_EXISTING_FEE_QUOTE_FEE_SUCCESS = "FETCH_EXISTING_FEE_QUOTE_FEE_SUCCESS";
var HANDLE_CHANGE_FROM_FORM = exports.HANDLE_CHANGE_FROM_FORM = "HANDLE_CHANGE_FROM_FORM";
var HANDLE_CHANGE_FROM_INPUT = exports.HANDLE_CHANGE_FROM_INPUT = "HANDLE_CHANGE_FROM_INPUT";
var HANDLE_BLUR_FROM_INPUT = exports.HANDLE_BLUR_FROM_INPUT = "HANDLE_BLUR_FROM_INPUT";
var DELETE_FEE = exports.DELETE_FEE = "DELETE_FEE";
var APPEND_FEE = exports.APPEND_FEE = "APPEND_FEE";
var REFRESH_FORM = exports.REFRESH_FORM = "REFRESH_FORM";
var CANCEL_EDITING = exports.CANCEL_EDITING = "CANCEL_EDITING";
var VALIDATE_QUOTE_FEES = exports.VALIDATE_QUOTE_FEES = "VALIDATE_QUOTE_FEES";
var SAVE_QUOTE_FEES = exports.SAVE_QUOTE_FEES = "SAVE_QUOTE_FEES";
