// @coxautokc/fusion-ui-components/lib/internal/CardDropdownMenu/CardDropdownMenu.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _DropdownButton = require('react-bootstrap/lib/DropdownButton');

var _DropdownButton2 = _interopRequireDefault(_DropdownButton);

var _MenuItem = require('react-bootstrap/lib/MenuItem');

var _MenuItem2 = _interopRequireDefault(_MenuItem);

var _DotsThreeVertical = require('react-entypo/lib/entypo/DotsThreeVertical');

var _DotsThreeVertical2 = _interopRequireDefault(_DotsThreeVertical);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var CardDropdownMenu = function CardDropdownMenu(_ref) {
  var htmlId = _ref.htmlId,
      actionDropdown = _ref.actionDropdown;

  var menuObj = actionDropdown || [];
  var menuItems = menuObj.map(function (v, key) {
    return _react2.default.createElement(
      _MenuItem2.default,
      {
        className: 'card__dropdown-item',
        key: key,
        eventKey: key,
        active: v.active,
        disabled: v.disabled,
        divider: v.divider,
        header: v.header,
        href: v.href,
        onClick: v.onClick,
        onSelect: v.onSelect
      },
      v.label
    );
  });
  var stopProp = function stopProp(e) {
    e.stopPropagation();
  };

  return _react2.default.createElement(
    'span',
    { id: htmlId },
    menuObj.length > 0 ? _react2.default.createElement(
      'div',
      { className: 'card__dropdown', onClick: stopProp },
      _react2.default.createElement(
        _DropdownButton2.default,
        {
          className: 'card__dropdown-button',
          bsStyle: 'link',
          title: _react2.default.createElement(_DotsThreeVertical2.default, { valign: true }),
          noCaret: true,
          pullRight: true,
          id: htmlId + '-CardDropdownButton'
        },
        menuItems
      )
    ) : null
  );
};

CardDropdownMenu.propTypes = {
  actionDropdown: _propTypes2.default.arrayOf(_propTypes2.default.shape({
    label: _propTypes2.default.string.isRequired,
    active: _propTypes2.default.boolean,
    disabled: _propTypes2.default.boolean,
    divider: _propTypes2.default.boolean,
    header: _propTypes2.default.boolean,
    href: _propTypes2.default.string,
    onClick: _propTypes2.default.func,
    onSelect: _propTypes2.default.func
  })).isRequired,
  htmlId: _propTypes2.default.string.isRequired
};

exports.default = CardDropdownMenu;
