// @coxautokc/fusion-ui-components/lib/ActionModal/ActionModal.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/** <ul>
 * <li>Action Modal serves as the wrapping component of the markup you wish to include in ActionBar's openToComponent Prop.</li>
 * <li>This is only used when you have a modal and will only accept Modal or FullPageModal as a direct child.</li>
 * <li>No visible example is needed this functionality is contained in ActionBar.</li>
 * </ul>
 */
var ActionModal = function (_Component) {
  _inherits(ActionModal, _Component);

  function ActionModal() {
    _classCallCheck(this, ActionModal);

    return _possibleConstructorReturn(this, (ActionModal.__proto__ || Object.getPrototypeOf(ActionModal)).apply(this, arguments));
  }

  _createClass(ActionModal, [{
    key: 'componentWillMount',
    value: function componentWillMount() {
      var children = this.props.children;

      if (Array.isArray(children)) {
        throw new Error('ActionModal only accepts 1 child');
      } else {
        return children;
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          htmlId = _props.htmlId;

      return _react2.default.createElement(
        'div',
        { id: htmlId },
        children
      );
    }
  }]);

  return ActionModal;
}(_react.Component);

ActionModal.propTypes = {
  /** Will only accept 1 child, and it must be a Modal or FullPageModal */
  children: _propTypes2.default.node.isRequired,

  /** Function used to close the modal. <br/><strong>You will not need to pass this, only receive it as ActionBar sets this.</strong>*/
  closeDialog: _propTypes2.default.func.isRequired,

  /** Globally unique and descriptive HTML ID. Used by QA for automated testing. */
  htmlId: _propTypes2.default.string,

  /** Set to true if modal is visible. <br/><strong>You will not need to pass this, only receive it as ActionBar sets this.</strong>*/
  showActionDialog: _propTypes2.default.bool.isRequired
};

ActionModal.defaultProps = {
  closeDialog: function closeDialog() {},
  htmlId: "",
  showActionDialog: false
};

exports.default = ActionModal;
