// @coxautokc/taxes/build/components/home-page.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Grid = require("react-bootstrap/lib/Grid");

var _Grid2 = _interopRequireDefault(_Grid);

var _Col = require("react-bootstrap/lib/Col");

var _Col2 = _interopRequireDefault(_Col);

var _FullPageModal = require("@coxautokc/fusion-ui-components/lib/FullPageModal");

var _FullPageModal2 = _interopRequireDefault(_FullPageModal);

var _ModalFooter = require("./customizedUiComponents/ModalFooter/ModalFooter");

var _ModalFooter2 = _interopRequireDefault(_ModalFooter);

require("font-awesome/css/font-awesome.css");

require("react-select/dist/react-select.css");

require("../index.scss");

require("@coxautokc/fusion-theme/dist/fusion-theme.min.css");

var _taxesPage = require("./taxes-page");

var _taxesPage2 = _interopRequireDefault(_taxesPage);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; } // eslint-disable-line


var HomePage = function (_Component) {
  _inherits(HomePage, _Component);

  function HomePage(props) {
    _classCallCheck(this, HomePage);

    var _this = _possibleConstructorReturn(this, (HomePage.__proto__ || Object.getPrototypeOf(HomePage)).call(this, props));

    _this.save = function () {
      var totalTaxAmount = "0";
      _this.props.populateValue(_this.props.name, totalTaxAmount);
      _this.props.closeDialog();
    };

    _this.state = {
      // showModal: false
    };
    return _this;
  }

  _createClass(HomePage, [{
    key: "render",
    value: function render() {
      var _props = this.props,
          isDisabled = _props.isDisabled,
          dialogStatus = _props.dialogStatus,
          closeDialog = _props.closeDialog;


      return _react2.default.createElement(
        "div",
        null,
        _react2.default.createElement(
          _FullPageModal2.default,
          {
            htmlId: "Taxes-Modal",
            className: "sample-fullpage-modal",
            show: dialogStatus,
            header: _react2.default.createElement(
              _FullPageModal2.default.Title,
              null,
              "EDIT TAXES"
            ),
            onHide: closeDialog,
            footer: _react2.default.createElement(_ModalFooter2.default, {
              displayResetBtn: true,
              modalName: "taxes",
              isDisabled: isDisabled,
              save: this.save,
              cancel: closeDialog,
              resetData: this.resetData
            })
          },
          _react2.default.createElement(
            _Grid2.default,
            {
              fluid: true,
              className: "template-page-wrapper page-wrapper__content-only"
            },
            _react2.default.createElement(
              _Col2.default,
              { sm: 12, md: 12, lg: 12 },
              _react2.default.createElement(_taxesPage2.default, {
                updateCurrentTab: this.props.updateCurrentTab,
                updateEditTaxProfileInput: this.props.updateEditTaxProfileInput,
                updateTaxSetupInput: this.props.updateTaxSetupInput,
                fetchTaxData: this.props.fetchTaxData,
                fetchFeeTaxData: this.props.fetchFeeTaxData,
                currentTab: this.props.currentTab,
                feeTaxSetup: this.props.feeTaxSetup,
                noTaxes: this.props.noTaxes,
                profile: this.props.profile,
                state: this.props.state,
                taxSetup: this.props.taxSetup,
                zipcode: this.props.zipcode
              })
            )
          )
        )
      );
    }
  }]);

  return HomePage;
}(_react.Component);

exports.default = HomePage;


HomePage.propTypes = {
  currentTab: _propTypes2.default.number,
  feeTaxSetup: _propTypes2.default.object,
  fetchFeeTaxData: _propTypes2.default.func,
  fetchTaxData: _propTypes2.default.func,
  isDisabled: _propTypes2.default.bool,
  closeDialog: _propTypes2.default.func.isRequired,
  dialogStatus: _propTypes2.default.bool.isRequired,
  populateValue: _propTypes2.default.func.isRequired,
  name: _propTypes2.default.string.isRequired,
  noTaxes: _propTypes2.default.bool,
  profile: _propTypes2.default.object,
  state: _propTypes2.default.string,
  taxSetup: _propTypes2.default.object,
  updateCurrentTab: _propTypes2.default.func,
  updateEditTaxProfileInput: _propTypes2.default.func,
  updateTaxSetupInput: _propTypes2.default.func,
  zipcode: _propTypes2.default.string
};
