// @coxautokc/taxes/build/components/taxes-page.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _editTaxProfile = require('./edit-tax-profile');

var _editTaxProfile2 = _interopRequireDefault(_editTaxProfile);

var _taxSetup = require('./tax-setup');

var _taxSetup2 = _interopRequireDefault(_taxSetup);

var _feeTaxSetup = require('./fee-tax-setup');

var _feeTaxSetup2 = _interopRequireDefault(_feeTaxSetup);

var _productCategoryTaxSetup = require('./product-category-tax-setup');

var _productCategoryTaxSetup2 = _interopRequireDefault(_productCategoryTaxSetup);

var _constants = require('../constants');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var TaxesPage = function (_React$Component) {
  _inherits(TaxesPage, _React$Component);

  function TaxesPage() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, TaxesPage);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = TaxesPage.__proto__ || Object.getPrototypeOf(TaxesPage)).call.apply(_ref, [this].concat(args))), _this), _this.getOnTabClickHandler = function (tab) {
      return function () {
        return _this.props.updateCurrentTab(tab);
      };
    }, _this.renderCurrentTab = function () {
      var content = _react2.default.createElement('div', null);
      switch (_this.props.currentTab) {
        case _constants.TaxesPageTabs["tax-setup"]:
          content = _react2.default.createElement(_taxSetup2.default, Object.assign({}, _this.props.taxSetup, {
            isDisabled: _this.props.noTaxes,
            updateInput: _this.props.updateTaxSetupInput }));
          break;
        case _constants.TaxesPageTabs["fee-tax-setup"]:
          content = _react2.default.createElement(_feeTaxSetup2.default, Object.assign({}, _this.props.feeTaxSetup, {
            isDisabled: _this.props.noTaxes }));
          break;
        case _constants.TaxesPageTabs["product-category-tax-setup"]:
          content = _react2.default.createElement(_productCategoryTaxSetup2.default, {
            isDisabled: _this.props.noTaxes });
          break;

        default:
          break;
      }
      return content;
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(TaxesPage, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.props.fetchTaxData();
      this.props.fetchFeeTaxData();
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          currentTab = _props.currentTab,
          noTaxes = _props.noTaxes,
          updateEditTaxProfileInput = _props.updateEditTaxProfileInput;

      return _react2.default.createElement(
        'div',
        { id: 'edit-taxes-page' },
        _react2.default.createElement(
          'div',
          null,
          _react2.default.createElement(_editTaxProfile2.default, Object.assign({}, this.props, {
            isDisabled: noTaxes,
            updateInput: updateEditTaxProfileInput }))
        ),
        _react2.default.createElement(
          'ul',
          { className: 'nav nav-tabs' },
          _react2.default.createElement(
            'li',
            { className: currentTab === _constants.TaxesPageTabs["tax-setup"] && 'active' },
            _react2.default.createElement(
              'a',
              { onClick: this.getOnTabClickHandler(_constants.TaxesPageTabs["tax-setup"]) },
              'Tax Setup '
            )
          ),
          _react2.default.createElement(
            'li',
            { className: currentTab === _constants.TaxesPageTabs["fee-tax-setup"] && 'active' },
            _react2.default.createElement(
              'a',
              { onClick: this.getOnTabClickHandler(_constants.TaxesPageTabs["fee-tax-setup"]) },
              'Fee tax Setup '
            )
          ),
          _react2.default.createElement(
            'li',
            { className: currentTab === _constants.TaxesPageTabs["product-category-tax-setup"] && 'active' },
            _react2.default.createElement(
              'a',
              { onClick: this.getOnTabClickHandler(_constants.TaxesPageTabs["product-category-tax-setup"]) },
              'Product category tax Setup'
            )
          )
        ),
        this.renderCurrentTab()
      );
    }
  }]);

  return TaxesPage;
}(_react2.default.Component);

exports.default = TaxesPage;


TaxesPage.propTypes = {
  currentTab: _propTypes2.default.number,
  feeTaxSetup: _propTypes2.default.object,
  fetchFeeTaxData: _propTypes2.default.func,
  fetchTaxData: _propTypes2.default.func,
  noTaxes: _propTypes2.default.bool,
  taxSetup: _propTypes2.default.object,
  updateCurrentTab: _propTypes2.default.func,
  updateEditTaxProfileInput: _propTypes2.default.func,
  updateTaxSetupInput: _propTypes2.default.func
};
