// @coxautokc/fusion-ui-components/lib/DatePicker/DatePicker.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _SingleDatePickerFieldContainer = require('./SingleDatePickerFieldContainer');

var _SingleDatePickerFieldContainer2 = _interopRequireDefault(_SingleDatePickerFieldContainer);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var DatePicker = function DatePicker(props) {
  return _react2.default.createElement(_SingleDatePickerFieldContainer2.default, props);
};

DatePicker.propTypes = {
  /** Additional class name. */
  className: _propTypes2.default.string,

  /** Globally unique and descriptive HTML ID. Used by QA for automated testing. */
  htmlId: _propTypes2.default.string.isRequired,

  /** Label text that displays above the date input text input. */
  label: _propTypes2.default.string.isRequired,

  /** onChange callback. Called with <code>cb(event, isValid)</code>. Check isValid to determine if the control is in error. */
  onDateChange: _propTypes2.default.func.isRequired,

  /** Mark label with asterisk and perform required field validation on blur if set to true. */
  required: _propTypes2.default.bool,

  /** Restricts past and unallowed dates if set to true. */
  restrictOutsideDays: _propTypes2.default.bool,

  /** Format of date expected from DatePicker.
   * <a href="https://momentjs.com/docs">Must use accepted moment.js formats.</a>
   */
  returnedDateFormat: _propTypes2.default.string,

  /** Initial date fed to DatePicker, accepts date object or string.</br>
   * Example objects:
   * <ul>
   * <li><code>moment('11-7-1980')</code></li>
   * <li><code>new Date('11-7-1980')</code></li>
   * </ul>
   * Example strings:
   * <ul>
   * <li><code>'11-7-1980'</code></li>
   * <li><code>'3/22/1980'</code></li>
   * <li><code>'Sat Mar 04 2017 02:58:34 GMT-0600 (Central Standard Time)'</code></li>
   * </ul>
   * Example ISO 8601 formatted strings:
   * <ul>
   * <li><code>'2020-12-03T01:30:43.000'</code></li>
   * <li>With timezone offset: <code>'2023-04-09T02:25:11.000-5:00'</code></li>
   * </ul>
  */
  selectedDate: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.object]),

  /** Format for date accepted in text input. <br />
   * Accepted formats:<br />
   * <ul>
   * <li>M/D/YY</li>
   * <li>M/D/YYYY</li>
   * <li>M-D-YY</li>
   * <li>M-D-YYYY</li>
   * <li>M.D.YY</li>
   * <li>M.D.YYYY</li>
   * <li>M D YY</li>
   * <li>M D YYYY</li>
   * </ul>
   */
  textFieldFormat: _propTypes2.default.string,

  /** Name associated with text input. */
  textFieldName: _propTypes2.default.string,

  /** Placeholder for text input. */
  textFieldPlaceholder: _propTypes2.default.string
};

exports.default = DatePicker;
