// @coxautokc/fusion-ui-components/lib/DatePicker/SingleDatePicker.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _DayPicker = require('react-dates/lib/components/DayPicker');

var _DayPicker2 = _interopRequireDefault(_DayPicker);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var SingleDatePicker = function SingleDatePicker(_ref) {
  var htmlId = _ref.htmlId,
      dayPickerContainerClasses = _ref.dayPickerContainerClasses,
      orientation = _ref.orientation,
      restrictOutsideDays = _ref.restrictOutsideDays,
      modifiers = _ref.modifiers,
      numberOfMonths = _ref.numberOfMonths,
      onDayMouseEnter = _ref.onDayMouseEnter,
      onDayMouseLeave = _ref.onDayMouseLeave,
      onDayClick = _ref.onDayClick,
      onPrevMonthClick = _ref.onPrevMonthClick,
      onNextMonthClick = _ref.onNextMonthClick,
      navPrev = _ref.navPrev,
      navNext = _ref.navNext,
      monthFormat = _ref.monthFormat,
      initialVisibleMonth = _ref.initialVisibleMonth,
      hideKeyboardShortcutsPanel = _ref.hideKeyboardShortcutsPanel,
      isRTL = _ref.isRTL,
      renderMonth = _ref.renderMonth;

  return _react2.default.createElement(
    'div',
    {
      id: htmlId,
      className: dayPickerContainerClasses
    },
    _react2.default.createElement(_DayPicker2.default, {
      orientation: orientation,
      enableOutsideDays: !restrictOutsideDays,
      modifiers: modifiers,
      numberOfMonths: numberOfMonths,
      onDayMouseEnter: onDayMouseEnter,
      onDayMouseLeave: onDayMouseLeave,
      onDayClick: onDayClick,
      onPrevMonthClick: onPrevMonthClick,
      onNextMonthClick: onNextMonthClick,
      navPrev: navPrev,
      navNext: navNext,
      monthFormat: monthFormat,
      hidden: false,
      initialVisibleMonth: initialVisibleMonth,
      hideKeyboardShortcutsPanel: hideKeyboardShortcutsPanel,
      isRTL: isRTL,
      renderMonth: renderMonth
    })
  );
};

// Components from 'react-dates'


SingleDatePicker.propTypes = {
  dayPickerContainerClasses: _propTypes2.default.string.isRequired,
  hideKeyboardShortcutsPanel: _propTypes2.default.bool,
  htmlId: _propTypes2.default.string.isRequired,
  initialVisibleMonth: _propTypes2.default.func,
  isRTL: _propTypes2.default.bool.isRequired,
  modifiers: _propTypes2.default.object.isRequired,
  monthFormat: _propTypes2.default.string.isRequired,
  navNext: _propTypes2.default.node,
  navPrev: _propTypes2.default.node,
  numberOfMonths: _propTypes2.default.number.isRequired,
  onDayClick: _propTypes2.default.func.isRequired,
  onDayMouseEnter: _propTypes2.default.func.isRequired,
  onDayMouseLeave: _propTypes2.default.func.isRequired,
  onNextMonthClick: _propTypes2.default.func,
  onPrevMonthClick: _propTypes2.default.func,
  orientation: _propTypes2.default.string.isRequired,
  renderMonth: _propTypes2.default.func,
  restrictOutsideDays: _propTypes2.default.bool.isRequired
};

exports.default = SingleDatePicker;
