// @coxautokc/fusion-ui-components/lib/DatePicker/SingleDatePickerField.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _FormControl = require('react-bootstrap/lib/FormControl');

var _FormControl2 = _interopRequireDefault(_FormControl);

var _FormGroup = require('react-bootstrap/lib/FormGroup');

var _FormGroup2 = _interopRequireDefault(_FormGroup);

var _InputGroup = require('react-bootstrap/lib/InputGroup');

var _InputGroup2 = _interopRequireDefault(_InputGroup);

var _HelpBlock = require('react-bootstrap/lib/HelpBlock');

var _HelpBlock2 = _interopRequireDefault(_HelpBlock);

var _Button = require('@coxautokc/fusion-ui-components/lib/Button');

var _Button2 = _interopRequireDefault(_Button);

var _Label = require('@coxautokc/fusion-ui-components/lib/internal/Label');

var _Label2 = _interopRequireDefault(_Label);

var _IconInsertInvitation = require('@coxautokc/fusion-ui-components/lib/Icons/IconInsertInvitation');

var _IconInsertInvitation2 = _interopRequireDefault(_IconInsertInvitation);

var _OutsideClickHandler = require('react-dates/lib/components/OutsideClickHandler');

var _OutsideClickHandler2 = _interopRequireDefault(_OutsideClickHandler);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var SingleDatePickerField = function SingleDatePickerField(_ref) {
  var className = _ref.className,
      datePicker = _ref.datePicker,
      dateString = _ref.dateString,
      displayLabel = _ref.displayLabel,
      error = _ref.error,
      htmlId = _ref.htmlId,
      label = _ref.label,
      onOutsideClick = _ref.onOutsideClick,
      onTextFieldKeyDown = _ref.onTextFieldKeyDown,
      onTextFieldChange = _ref.onTextFieldChange,
      onTextFieldFocus = _ref.onTextFieldFocus,
      onToggleInputButton = _ref.onToggleInputButton,
      textFieldName = _ref.textFieldName,
      textFieldPlaceholder = _ref.textFieldPlaceholder,
      required = _ref.required,
      warning = _ref.warning;

  var isError = !!error;
  var isWarning = !!warning && !isError;

  return _react2.default.createElement(
    'div',
    { id: htmlId, className: "SingleDatePicker " + className },
    _react2.default.createElement(
      _OutsideClickHandler2.default,
      { onOutsideClick: onOutsideClick },
      _react2.default.createElement(
        _FormGroup2.default,
        {
          controlId: 'datePickerTextField',
          className: 'datePickerTextField__formGroup',
          validationState: isError ? 'error' : isWarning ? 'warning' : null
        },
        ' ',
        displayLabel && _react2.default.createElement(
          _Label2.default,
          {
            className: className,
            required: required,
            htmlId: htmlId + '-label'
          },
          label
        ),
        _react2.default.createElement(
          _InputGroup2.default,
          {
            onKeyDown: onTextFieldKeyDown
          },
          _react2.default.createElement(_FormControl2.default, {
            name: textFieldName,
            onChange: onTextFieldChange,
            onClick: onTextFieldFocus,
            onFocus: onTextFieldFocus,
            placeholder: textFieldPlaceholder,
            type: 'text',
            value: dateString
          }),
          _react2.default.createElement(
            _InputGroup2.default.Button,
            null,
            _react2.default.createElement(
              _Button2.default,
              {
                htmlId: htmlId + '-toggleButton',
                onClick: onToggleInputButton,
                tabIndex: -1,
                className: "SingleDatePicker__trigger-button",
                buttonStyle: isError ? 'danger' : 'default'
              },
              _react2.default.createElement(_IconInsertInvitation2.default, null)
            )
          )
        ),
        isError ? _react2.default.createElement(
          _HelpBlock2.default,
          null,
          error
        ) : isWarning ? _react2.default.createElement(
          _HelpBlock2.default,
          null,
          warning
        ) : null
      ),
      datePicker
    )
  );
};

// Components from 'react-dates'


SingleDatePickerField.propTypes = {
  className: _propTypes2.default.string,
  datePicker: _propTypes2.default.object,
  dateString: _propTypes2.default.string.isRequired,
  displayLabel: _propTypes2.default.bool,
  error: _propTypes2.default.string.isRequired,
  htmlId: _propTypes2.default.string.isRequired,
  label: _propTypes2.default.string.isRequired,
  onOutsideClick: _propTypes2.default.func.isRequired,
  onTextFieldChange: _propTypes2.default.func.isRequired,
  onTextFieldFocus: _propTypes2.default.func.isRequired,
  onTextFieldKeyDown: _propTypes2.default.func.isRequired,
  onToggleInputButton: _propTypes2.default.func.isRequired,
  required: _propTypes2.default.bool.isRequired,
  textFieldName: _propTypes2.default.string.isRequired,
  textFieldPlaceholder: _propTypes2.default.string.isRequired,
  warning: _propTypes2.default.string.isRequired
};

exports.default = SingleDatePickerField;
