// @coxautokc/taxes/build/containers/home-page.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _reactRedux = require("react-redux");

var _extend = require("lodash/extend");

var _extend2 = _interopRequireDefault(_extend);

var _taxesPageActions = require("../actions/taxes-page-actions");

var _homePage = require("../components/home-page");

var _homePage2 = _interopRequireDefault(_homePage);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function mapStateToProps(state) {
  return (0, _extend2.default)({}, {
    currentTab: state.taxesPage.currentTab,
    feeTaxSetup: state.taxesPage.feeTaxSetup,
    noTaxes: state.taxesPage.noTaxes,
    profile: state.taxesPage.profile,
    state: state.taxesPage.state,
    taxSetup: state.taxesPage.taxSetup,
    zipcode: state.taxesPage.zipcode
  });
}

function mapDispatchToProps(dispatch) {
  return {
    updateCurrentTab: function updateCurrentTab(tab) {
      return dispatch((0, _taxesPageActions.updateTaxesPageCurrentTab)(tab));
    },
    updateEditTaxProfileInput: function updateEditTaxProfileInput(input, value) {
      return dispatch((0, _taxesPageActions.updateEditTaxProfileInput)(input, value));
    },
    updateTaxSetupInput: function updateTaxSetupInput(input, value) {
      return dispatch((0, _taxesPageActions.updateTaxSetupInput)(input, value));
    },
    fetchTaxData: function fetchTaxData() {
      return dispatch((0, _taxesPageActions.fetchTaxSetupData)());
    },
    fetchFeeTaxData: function fetchFeeTaxData() {
      return dispatch((0, _taxesPageActions.fetchFeeTaxSetupTableData)());
    }
  };
}

function mergeProps(stateProps, dispatchProps, ownProps) {
  return (0, _extend2.default)({}, stateProps, dispatchProps, ownProps);
}

exports.default = (0, _reactRedux.connect)(mapStateToProps, mapDispatchToProps, mergeProps)(_homePage2.default);
